### ISSUE TITLE
New Component: DynamicVariableSpaceProxy

### ISSUE THREAD

**Issue by JackTheFoxOtter** [Yellow-Dog-Man/Resonite-Issues#1250]
### Is your feature request related to a problem? Please describe.

Dynamic variables are very useful for storing information in specific slot hierarchies, and I use them in pretty much every project that requires more complex data storage. Isolating them to only a certain slot hierarchy is what makes them so versatile, however, sometimes you run into situations where you need to dynamic information accross hierarchies, and that's when stuff usually starts getting a bit messy.

### Describe the solution you'd like

I would like a DynamicVariableSpaceProxy component, that dynamic variables can bind to, but only actually proxies an existing DynamicVariableSpace component, which could be in a different hierarchy. That way, you can link up different hierarchies to share the same dynamic variables, allowing for an elegant solution to connect them up.

### Describe alternatives you've considered

Either manually copying the information, driving them from one hierarchy to another, or using dynamic fields to link them up. All of those have different applications, but the one thing they have in common is that you always need to link up every single value you need synced individually, which from my experience results in a lot of extra clutter in your systems.

### Additional Context

For me, the main use for this is whenever you have a central information storage that you want to link up to UI elements in a different hierarchy. While usually I split those up for organization, there are some cases in which you have to have them split into separate hierarchies, for example when creating facets and working with the AlwaysOnFacetModule, which will split your creation up into two hierarchies, or when your UI elements represent a list of information, where you can't have them link up to the same dynamic variable space to begin with. I imagine this would also be very handy for the upcoming data feed / UI instantiation system.

**Comment by BaxterOttoman** [comment]
I can't speak to Jack's needs, but I can say that I can easily see situations where you want to have multiple objects under a specific parent that you would like to individually target other hierarchies. As a simple example, creating UIX elements that you want to take data from various targets. Being able to drive the contents of those UIX elements with data from slots located elsewhere could streamline making interfaces for some workflows.

**Comment by JackTheFoxOtter** [comment]
> I'm not sure I'm understanding this issue, @JackTheFoxOtter - why would you not just place the variables that need to be shared across multiple hierarchies in a singular slot under which the other hierarchies are children of?
> 
> Or making use of the "World" variable space?

@shiftyscales That does not work in many cases. As per one of the examples I've brought up, the facet with AlwaysOnFacetModule for example, it's impossible to keep that creation in a shared hierarchy with a custom DynamicVariableSpace. But even in more trivial cases, whenever you have a "template" with a DynamicVariableSpace, and you duplicate that template, you now have two DynamicVariableSpaces, each with unique values. If you now want to make UI that interfaces with one or multiple of those templates, you'll run into the issues I outlined regarding dynamically binding that UI (or controller / modular system etc.) to the individual instance. You can't just put the DynamicVariableSpace component into the parent of all instances, because that would mean the values collide - for example, every instance that has a string variable called "Name" now exists in the same DynamicVariableSpace, meaning only one of the names will actually stay and the rest get overwritten with it.

With the Proxy component, if it has "Target" field with a DynamicVariableSpace, you could write to that field to dynamically change the space the component is proxying, which would present a pretty elegant solution for this problem in my opinion.

There are many situations in which you need to have separate DynamicVariableSpaces. Currently, there exists no elegant solution to dynamically link those entirely to other parts of the system - that's why I proposed this feature.

**Comment by JackTheFoxOtter** [comment]
To add even more context here - the dynamic linking stuff is just my specific use-case, but I can see how this would be valuable in many more situations as well. Any system that has a sort of "controller" object and multiple instances that might not be in the same hierarchy because they reparent or can be picked up could be simplified with this. It's not gonna allow something that wasn't possible before, but it would massively simplify the alternative solutions and make dynamic variables even more versatile.

**Comment by JackTheFoxOtter** [comment]
> If potential conflicts are an issue- that is why it is recommended/considered best practice to have high specificity in variable names- same for anything else in the engine that works similarly, e.g. snappers.
> 
> I don't see why using the world space would be something to be avoided- why do you believe that to be the case?

You can put stuff on the world space if it makes sense, like, you have objects that are spawned independently from one another and need to find each other, or you want arbitrary objects to react to some standardized world variable, like systems that provide a world "beat" or color.

But if your variables only concern your creation, you should definitely make a specific DynamicVariableSpace for it that only affects your system. As a general design consideration, before it makes no sense to have something "publicly accessible" that you only expect your single creation to affect, but also for performance to some degree.

It's also not guaranteed to be there, someone could just remove it. Generally, in most cases, you don't want to have your objects state / variables stored outside of that object.

**Comment by gentlecolts** [comment]
I've asked for something like this before #111

**Comment by JackTheFoxOtter** [comment]
I don't think this is the same request.

The linked request is similar, but it suggested adding a proxy component that targets a different hierarchy than the one it is attached to - which Froox pointed out as not supported by the DataModel.

My suggestion was adding a proxy component that behaves exactly like the normal DynamicVariableSpace, so only affecting the hierarchy it is attached to, with the difference that it doesn't have its own "dictionary" to store the variables in, instead using one of an existing dynamic variable space. From my understanding, that should be possible. If it isn't, I'd prefer an explanation as to why.

**Comment by JackTheFoxOtter** [comment]
What's the correct process to "appealing" a ticket closure? I'm worried that because it's closed now it's no longer on anyone's radar to check comments and potentially reopen it.

**Comment by JackTheFoxOtter** [comment]
Yes. My idea would have been to just have field "Target" of type DynamicVariableSpace on the DynamicVariableSpaceProxy. That way you could easily change that in ProtoFlux, and there is no ambiguity about binding to space names, as those aren't even relevant - what matters is the component that contains the dictionary.

**Comment by JackTheFoxOtter** [comment]
The one thing that might require considering is what to do with the variables when the target changes. Would that count as them rebinding, so that which value takes authority in case of a conflict depends on the OverwriteOnLink option per variable?

**Comment by Frooxius** [comment]
Yes, that would force the variables to be rebound. It would behave same as if you moved them in the hierarchy under a different space.

**Comment by mpmxyz** [comment]
I played around with this idea in my head and found a use-case for the proxy space to have its own name instead of implicitly inheriting it from the connected space (or whatever alternative):

Imagine a sales receipt that represents a connection between two users.
Both users would have spaces with the name "User".
If the relay had its own name you could access one user's space with "Seller" and the other one with "Buyer".

Apart from change tracking you also win slightly less verbose ProtoFlux:
- `DynamicVariableInput("Buyer/Cash")` instead of `ReadDynamicVariable(DynamicVariableInput("Buyer"), "User/Cash")`
- `WriteDynamicVariable("Buyer/Cash", 0)` instead of `WriteDynamicVariable(DynamicVariableInput("Buyer"), "User/Cash", 0)`
